src/server/method_callback.rs

```diff
@@ -0,0 +1,215 @@
+use ::core::ffi::c_void;
+use std::{
+    panic::{catch_unwind, AssertUnwindSafe},
+    ptr::NonNull,
+};
+
+use open62541_sys::{
+    UA_MethodCallback, UA_NodeId, UA_Server, UA_StatusCode, UA_Variant, UA_EMPTY_ARRAY_SENTINEL,
+};
+use thiserror::Error;
+
+use crate::{server::NodeContext, ua, DataType as _};
+
+/// Result from [`MethodCallback`] operations.
+///
+/// On success, the operations return `Ok(())`. The actual value is transmitted through the
+/// `context` argument. See [`MethodCallback::call()`] for details.
+#[allow(clippy::module_name_repetitions)]
+pub type MethodCallbackResult = Result<(), MethodCallbackError>;
+
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Error)]
+pub enum MethodCallbackError {
+    #[error("{0}")]
+    StatusCode(ua::StatusCode),
+
+    #[error("not supported")]
+    NotSupported,
+}
+
+impl MethodCallbackError {
+    pub(crate) fn into_status_code(self) -> ua::StatusCode {
+        match self {
+            MethodCallbackError::StatusCode(status_code) => status_code,
+            MethodCallbackError::NotSupported => ua::StatusCode::BADNOTSUPPORTED,
+        }
+    }
+}
+
+impl From<ua::StatusCode> for MethodCallbackError {
```

Comment:
Not any status code is a `MethodCallbackError` error. This conversions should become an explicit method `from_status_code`. Such a conversion might be convenient for using the `?` operator but is semantically wrong.

Reply:
What is `MethodCallbackError`? Just a type-safe alias for `ua::StatusCode` on the same abstraction level?

From my understandung `ua::StatusCode` is more versatile and less specific. It's meaning depend on the context. 

The `From/Into` trait would allow to unintentionally convert ANY `ua::StatusCode` that occurred outside the _Method Context_ context into a `MethodCallbackError`. You should better not expose such a feature in your public API.

`DataSourceError` seems to do this correctly by offering explicit conversion methods.